File: ios/Sources/RynBridge/MessageDeserializer.swift
import Foundation

public struct MessageDeserializer: Sendable {
    private let decoder: JSONDecoder

    public init() {
        self.decoder = JSONDecoder()
    }

    public func deserialize(_ raw: String) throws -> IncomingMessage {
        guard let data = raw.data(using: .utf8) else {
            throw RynBridgeError(code: .invalidMessage, message: "Failed to convert message to UTF-8 data")
        }

        let json: [String: AnyCodable]
        do {
            json = try decoder.decode([String: AnyCodable].self, from: data)
        } catch {
            throw RynBridgeError(code: .invalidMessage, message: "Failed to parse JSON: \(error.localizedDescription)")
        }

        guard let id = json["id"]?.stringValue, !id.isEmpty else {
            throw RynBridgeError(code: .invalidMessage, message: "Message missing required field: id")
        }

        // Discriminate by "status" field: present → response, absent → request
        if json["status"] != nil {
            let response = try decodeResponse(from: data)
            return .response(response)
        } else {
            let request = try decodeRequest(from: data)
            return .request(request)
        }
    }

    private func decodeRequest(from data: Data) throws -> BridgeRequest {
        do {
            let request = try decoder.decode(BridgeRequest.self, from: data)
            guard !request.module.isEmpty else {
                throw RynBridgeError(code: .invalidMessage, message: "Request missing required field: module")
            }
            guard !request.action.isEmpty else {
                throw RynBridgeError(code: .invalidMessage, message: "Request missing required field: action")
            }
            return request
        } catch let error as RynBridgeError {
            throw error
        } catch {
            throw RynBridgeError(code: .invalidMessage, message: "Failed to decode request: \(error.localizedDescription)")
        }
    }

    private func decodeResponse(from data: Data) throws -> BridgeResponse {
        do {
            return try decoder.decode(BridgeResponse.self, from: data)
        } catch {
            throw RynBridgeError(code: .invalidMessage, message: "Failed to decode response: \(error.localizedDescription)")
        }
    }
}

File: android/core/src/main/kotlin/com/devryner/rynbridge/core/MessageDeserializer.kt
package com.devryner.rynbridge.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class MessageDeserializer {

    private val json = Json { ignoreUnknownKeys = true }

    fun deserialize(raw: String): IncomingMessage {
        val jsonElement = try {
            json.parseToJsonElement(raw)
        } catch (e: Exception) {
            throw RynBridgeError(
                code = ErrorCode.INVALID_MESSAGE,
                message = "Failed to parse JSON: ${e.message}"
            )
        }

        val obj = jsonElement.jsonObject

        val id = obj["id"]?.jsonPrimitive?.content
        if (id.isNullOrEmpty()) {
            throw RynBridgeError(
                code = ErrorCode.INVALID_MESSAGE,
                message = "Message missing required field: id"
            )
        }

        // Discriminate by "status" field: present → response, absent → request
        return if (obj.containsKey("status")) {
            val response = decodeResponse(raw)
            IncomingMessage.Response(response)
        } else {
            val request = decodeRequest(raw)
            IncomingMessage.Request(request)
        }
    }

    private fun decodeRequest(raw: String): BridgeRequest {
        val request = try {
            json.decodeFromString<BridgeRequest>(raw)
        } catch (e: RynBridgeError) {
            throw e
        } catch (e: Exception) {
            throw RynBridgeError(
                code = ErrorCode.INVALID_MESSAGE,
                message = "Failed to decode request: ${e.message}"
            )
        }

        if (request.module.isEmpty()) {
            throw RynBridgeError(
                code = ErrorCode.INVALID_MESSAGE,
                message = "Request missing required field: module"
            )
        }
        if (request.action.isEmpty()) {
            throw RynBridgeError(
                code = ErrorCode.INVALID_MESSAGE,
                message = "Request missing required field: action"
            )
        }

        return request
    }

    private fun decodeResponse(raw: String): BridgeResponse {
        try {
            return json.decodeFromString<BridgeResponse>(raw)
        } catch (e: Exception) {
            throw RynBridgeError(
                code = ErrorCode.INVALID_MESSAGE,
                message = "Failed to decode response: ${e.message}"
            )
        }
    }
}
